File: backend/app/schemas/idc.py
"""数据中心 (IDC) Pydantic Schema (phase: datacenter)。

与前端数据中心管理/切换/对比一致:
- 列表/详情返回 camelCase 字段, is_current 标识当前默认中心。
- compare 返回各中心电力/制冷/机柜/告警的并排对比指标。
- alarms 返回跨中心统一告警汇总。
"""
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class IdcBase(BaseModel):
    code: str = Field(..., max_length=32, description="站点编码 如 EC1-HZ (唯一)")
    name: str = Field(..., max_length=128, description="数据中心名称")
    region: str = Field("", max_length=64, description="地域/可用区")
    address: str = Field("", max_length=255, description="地址")
    power_capacity_mw: float = Field(0, description="电力容量 MW")
    cooling_capacity_mw: float = Field(0, description="制冷容量 MW")
    rack_capacity: int = Field(0, description="机柜总容量")
    rooms: int = Field(0, description="包间数量")
    status: str = Field("运营", max_length=16, description="运营/建设/下线")
    capacity_kw: int = Field(0, description="机柜额定功率 kW (单体)")
    description: str = Field("", max_length=512, description="站点说明")


class IdcCreate(IdcBase):
    pass


class IdcUpdate(BaseModel):
    code: Optional[str] = None
    name: Optional[str] = None
    region: Optional[str] = None
    address: Optional[str] = None
    power_capacity_mw: Optional[float] = None
    cooling_capacity_mw: Optional[float] = None
    rack_capacity: Optional[int] = None
    rooms: Optional[int] = None
    status: Optional[str] = None
    capacity_kw: Optional[int] = None
    description: Optional[str] = None


class IdcOut(IdcBase):
    id: int
    is_current: bool = False
    created_at: str = ""
    updated_at: str = ""


class IdcCompareMetric(BaseModel):
    """单个中心的对比指标行。"""
    id: int
    code: str
    name: str
    region: str
    status: str
    power_capacity_mw: float
    cooling_capacity_mw: float
    rack_capacity: int
    rack_used: int = 0
    device_count: int = 0
    online_count: int = 0
    active_alarm_count: int = 0


class IdcCompareOut(BaseModel):
    centers: list[IdcCompareMetric] = Field(default_factory=list)
    current_idc_id: Optional[int] = None


class IdcAlarmItem(BaseModel):
    idc_id: int
    idc_name: str
    idc_code: str
    alarm_id: Optional[str] = None
    device_id: Optional[str] = None
    category: Optional[str] = None
    metric_name: Optional[str] = None
    level: Optional[str] = None
    value: Optional[float] = None
    unit: Optional[str] = None
    desc: Optional[str] = None
    state: Optional[str] = None
    ts: Optional[float] = None


class IdcAlarmsOut(BaseModel):
    total: int
    items: list[IdcAlarmItem] = Field(default_factory=list)
    by_idc: dict[int, int] = Field(default_factory=dict)

File: frontend/src/api/idc.ts
import request from './request'

export interface Idc {
  id: number
  code: string
  name: string
  region: string
  address: string
  powerCapacityMw: number
  coolingCapacityMw: number
  rackCapacity: number
  rooms: number
  status: string
  capacityKw: number
  description: string
  isCurrent: boolean
  createdAt?: string
  updatedAt?: string
}

export interface IdcCreate {
  code: string
  name: string
  region: string
  address: string
  powerCapacityMw: number
  coolingCapacityMw: number
  rackCapacity: number
  rooms: number
  status: string
  capacityKw: number
  description: string
}

export interface IdcCompareItem {
  id: number
  code: string
  name: string
  region: string
  status: string
  powerCapacityMw: number
  coolingCapacityMw: number
  rackCapacity: number
  rackUsed: number
  deviceCount: number
  onlineCount: number
  activeAlarmCount: number
}

export interface IdcCompare {
  centers: IdcCompareItem[]
  currentIdcId: number | null
}

export interface IdcAlarm {
  idcId: number
  idcName: string
  idcCode: string
  alarmId?: string
  deviceId?: string
  category?: string
  metricName?: string
  level?: string
  value?: number
  unit?: string
  desc?: string
  state?: string
  ts?: number
}

export interface IdcAlarmResp {
  total: number
  items: IdcAlarm[]
  byIdc: Record<string, number>
}

export interface IdcRelatedService {
  key: string
  name: string
  deviceCount: number
  onlineCount: number
  alarmCount: number
}

export interface IdcServicesResp {
  idcId: number
  idcName: string
  services: IdcRelatedService[]
  totalDevices: number
  onlineDevices: number
}

export interface IdcOpLog {
  id: number
  ts: string
  action: string
  target: string
  operator: string
  detail: string
}

export interface IdcOpLogsResp {
  total: number
  items: IdcOpLog[]
}

export interface IdcBatchDeleteResp {
  deleted: number
  skipped: number[]
}

export interface IdcToggleStatusResp {
  id: number
  status: string
  isCurrent: boolean
}

export const listIdcs = (params: { region?: string; status?: string } = {}) =>
  request.get<unknown, Idc[]>('/api/idc', { params })

export const getCurrentIdc = () => request.get<unknown, Idc>('/api/idc/current')

export const setIdcCurrent = (id: number) =>
  request.put<unknown, Idc>(`/api/idc/current?idcId=${id}`, {})

export const compareIdcs = () => request.get<unknown, IdcCompare>('/api/idc/compare')

export const unifiedAlarms = () => request.get<unknown, IdcAlarmResp>('/api/idc/alarms')

export const createIdc = (data: IdcCreate) =>
  request.post<unknown, Idc>('/api/idc', data)

export const updateIdc = (id: number, data: Partial<IdcCreate>) =>
  request.put<unknown, Idc>(`/api/idc/${id}`, data)

export const deleteIdc = (id: number) => request.delete(`/api/idc/${id}`)

export const batchDeleteIdcs = (ids: number[]) =>
  request.post<unknown, IdcBatchDeleteResp>('/api/idc/batch-delete', { ids })

export const toggleIdcStatus = (id: number) =>
  request.put<unknown, IdcToggleStatusResp>(`/api/idc/${id}/toggle-status`, {})

export const getIdcServices = (id: number) =>
  request.get<unknown, IdcServicesResp>(`/api/idc/${id}/services`)

export const getIdcOpLogs = (limit = 50) =>
  request.get<unknown, IdcOpLogsResp>('/api/idc/op-logs', { params: { limit } })
